File: clevertap-core/src/main/java/com/clevertap/android/sdk/response/NdFcapGate.java
package com.clevertap.android.sdk.response;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.clevertap.android.sdk.Constants;
import com.clevertap.android.sdk.Logger;
import com.clevertap.android.sdk.NdFCManager;
import com.clevertap.android.sdk.displayunits.model.CleverTapDisplayUnit;

import org.json.JSONObject;

import java.util.ArrayList;

/**
 * Native Display frequency-cap gate (SDK-6055). Filters display units through
 * {@link NdFCManager#canShow} before they are handed to the host — the ND analog of in-app's
 * pre-display cap gate.
 *
 * <p>Only units carrying an fcap marker ({@code efc}/{@code tlc}/{@code tdc}/{@code mdc}/
 * {@code excludeGlobalFCaps}) are gated; unmarked units pass through unchanged so existing non-fcap
 * display units are never affected. Advanced {@code frequencyLimits}/{@code occurrenceLimits} were
 * already applied during evaluation (the server only ships content for {@code adUnit_eval}-voted
 * campaigns), so the delivery-time re-check here covers the counter caps only.
 */
final class NdFcapGate {

    private NdFcapGate() {
    }

    @NonNull
    static ArrayList<CleverTapDisplayUnit> filter(
            @NonNull ArrayList<CleverTapDisplayUnit> units,
            @Nullable NdFCManager ndFCManager,
            @NonNull Logger logger,
            String accountId
    ) {
        if (ndFCManager == null) {
            return units;
        }
        final ArrayList<CleverTapDisplayUnit> allowed = new ArrayList<>(units.size());
        for (CleverTapDisplayUnit unit : units) {
            final JSONObject json = unit.getJsonObject();
            if (json == null || !isFcapManaged(json)) {
                allowed.add(unit); // not fcap-managed -> deliver as before
                continue;
            }
            final boolean excludeFromCaps =
                    json.optInt(Constants.KEY_EFC, -1) == 1
                            || json.optInt(Constants.KEY_EXCLUDE_GLOBAL_CAPS, -1) == 1;
            final boolean canShow = ndFCManager.canShow(
                    unit.getUnitID(),
                    excludeFromCaps,
                    json.optInt(Constants.KEY_TLC, -1),   // -1 = uncapped
                    json.optInt(Constants.KEY_TDC, -1),   // -1 = uncapped
                    json.optInt(Constants.INAPP_MAX_DISPLAY_COUNT, -1),
                    false /* advanced whenLimits already applied at evaluation time */);
            if (canShow) {
                allowed.add(unit);
            } else {
                logger.verbose(accountId,
                        Constants.FEATURE_DISPLAY_UNIT + "ND unit " + unit.getUnitID()
                                + " suppressed by frequency caps");
            }
        }
        return allowed;
    }

    private static boolean isFcapManaged(@NonNull JSONObject json) {
        return json.has(Constants.KEY_EFC)
                || json.has(Constants.KEY_TLC)
                || json.has(Constants.KEY_TDC)
                || json.has(Constants.INAPP_MAX_DISPLAY_COUNT)
                || json.has(Constants.KEY_EXCLUDE_GLOBAL_CAPS);
    }
}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/response/NdFcapGate.kt
package com.clevertap.android.sdk.response

import com.clevertap.android.sdk.Constants
import com.clevertap.android.sdk.ILogger
import com.clevertap.android.sdk.NdFCManager
import com.clevertap.android.sdk.displayunits.model.CleverTapDisplayUnit

/**
 * Native Display frequency-cap gate. Filters display units through [NdFCManager.canShow]
 * before they are handed to the host — the ND analog of in-app's pre-display cap gate.
 *
 * Only units carrying an fcap marker (`efc`/`tlc`/`tdc`/`mdc`/`excludeGlobalFCaps`) are gated; unmarked
 * units pass through unchanged so existing non-fcap display units are never affected. Advanced
 * `frequencyLimits`/`occurrenceLimits` were already applied during evaluation (the server only ships
 * content for `adUnit_eval`-voted campaigns), so the delivery-time re-check here covers counter caps only.
 */
internal object NdFcapGate {

    private const val UNCAPPED = -1

    fun filter(
        units: List<CleverTapDisplayUnit>,
        ndFCManager: NdFCManager?,
        logger: ILogger,
        accountId: String?,
    ): List<CleverTapDisplayUnit> {
        if (ndFCManager == null) {
            return units
        }
        return units.filter { unit ->
            val json = unit.jsonObject
            if (json == null || !NdFCManager.isFcapManaged(json)) {
                return@filter true // not fcap-managed -> deliver as before
            }
            val excludeFromCaps = json.optInt(Constants.KEY_EFC, UNCAPPED) == 1 ||
                json.optInt(Constants.KEY_EXCLUDE_GLOBAL_CAPS, UNCAPPED) == 1
            // Gate on the stable campaign id (ti), NOT unitID (= wzrk_id, ti_yyyyMMdd, rotates daily) —
            // must match how didShow records and how the evaluator keys whenLimits.
            val campaignId = json.optString(Constants.INAPP_ID_IN_PAYLOAD)
            val canShow = ndFCManager.canShow(
                campaignId,
                excludeFromCaps,
                json.optInt(Constants.KEY_TLC, UNCAPPED),
                json.optInt(Constants.KEY_TDC, UNCAPPED),
                json.optInt(Constants.INAPP_MAX_DISPLAY_COUNT, UNCAPPED),
                false, // advanced whenLimits already applied at evaluation time
            )
            if (!canShow) {
                logger.verbose(
                    accountId,
                    "${Constants.FEATURE_DISPLAY_UNIT}ND unit ${unit.unitID} suppressed by frequency caps",
                )
            }
            canShow
        }
    }
}
